keps/sig-auth/3299-kms-v2-improvements/README.md

```diff
@@ -0,0 +1,524 @@
+
+**Health Check & Status**: Today, the health check from kube-apiserver to KMS plugin is an `Encrypt` operation followed by `Decrypt` operation. These operations cost money in cloud environments and do not allow the plugin to perform more holistic checks on if it is healthy. Furthermore, a plugin has no way to inform the API server if its underlying key encryption key (KEK) has been rotated. If we provide a separate status RPC call with its own `StatusRequest` and `StatusResponse`, the KMS plugin can indicate the change in KEK version as part of response. This could be an indication that the KEK is now rotated and storage migration is now required.
+
+**Observability**: The only way to correlate a successful/failed envelope operation today is to use the approximate timestamp of the operation to check events in kube-apiserver, kms-plugin and KMS. There is no guarantee that the timestamp of the operation is the same as the timestamp of the corresponding event in KMS. This KEP proposes extending the signature of the kms-plugin interface to include the transaction ID (to be generated by the kube-apiserver), which kms-plugin could pass to KMS. This transaction ID will be logged in the kube-apiserver with additional metadata such as secret name and namespace for the envelope operation. Similarly, the transaction ID will be logged in the kms-plugin and optionally passed to KMS.
+
+### Goals
+- improve readiness times for clusters with a large number of encrypted resources
+- reduce the likelihood of hitting the KMS rate limit
+- enable fully automated key rotation for the latest key
+- improve KMS plugin health check reliability
+- improve observability of envelop operations between kube-apiserver, KMS plugins and KMS
+- if this v2 API reaches GA in release N, the existing v1beta1 gRPC API will be deprecated at release N and removed at release N+3 (the existing key rotation dance of using multiple providers will be used to migrate from v1beta1 to v2)
+
+### Non-Goals
+- Prevent KMS rate limiting
+- Recovery when KMS KEK is deleted
+- Using the proposed transaction ID for audit logging
+
+## Proposal
+
+Performance, Health Check, Observability and Rotation:
+- Support key hierarchy in KMS plugin that generates local KEK
```

Comment:
What prevents plugins from using key hierarchies today?

Reply:
I misspoke on the last SIG auth call so I will try to provide correct details here.

I do not think anything explicitly prevents the existing KMS plugins from using a key hierarchy today.  However, we do not encourage/enable it either.  Specifically, a v1beta1 plugin would (assuming it does not want to use some external database) need to encode the key hierarchy into the eDEK response (which has an undocumented length limit of 65kb).  This is certainly possible, but hardly ideal.  This leads to the data stored on disk being a mixture of the bespoke encoding format that Kube uses and whatever format the plugin comes up with to encode structure into the eDEK.  I think we can do a lot better, and it is much easier and safer to make these large sweeping changes in a new API (especially now that we have a much stronger process around enhancements/PRR).  Since a new API doesn't need to be backward compatible, we can require that plugins implement and understand the new endpoints (and we can keep supporting both APIs for some time to allow for data migration).  Having a dedicated and consistent structured format for storage will also make it possible to write a single tool to introspect encrypted data (for example, to determine what external KMS keys are currently in-use on a given cluster).

Reply:
Ok, thanks for clarifying.